feat(http): better binary support & XHR support

charsleysa opened the pull request:
<!--
We, the rest of the NativeScript community, thank you for your
contribution! 
To help the rest of the community review your change, please follow the instructions in the template.
-->

<!-- PULL REQUEST TEMPLATE -->
<!-- (Update "[ ]" to "[x]" to check a box) -->

## PR Checklist

- [x] The PR title follows our guidelines: https://github.com/NativeScript/NativeScript/blob/master/CONTRIBUTING.md#commit-messages.
- [x] There is an issue for the bug/feature this PR is for. To avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it.
- [x] You have signed the [CLA](http://www.nativescript.org/cla).
- [x] All existing tests are passing: https://github.com/NativeScript/NativeScript/blob/master/DevelopmentWorkflow.md#running-unit-tests.
- [x] Tests for the changes are included - https://github.com/NativeScript/NativeScript/blob/master/WritingUnitTests.md.

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
HTTP module does not support uploading binary information, only text based information.
XHR doesn't follow spec.

## What is the new behavior?
<!-- Describe the changes. -->
HTTP module now supports uploading binary information through the use of ArrayBuffer.
XHR now better follows spec, now has support for Blob and ArrayBuffer.

Implements #570.
Closes #7632.
Closes NativeScript/nativescript-angular#1202.

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

BREAKING CHANGES:


Possibly a breaking change if anyone is using the `org.nativescript.widgets2.Async.Http` Java class directly as `RequestOptions.content` has changed from a type `String` to type `java.nio.ByteBuffer`, though I'm not sure if this is considered to be an internal API.

Also, XHR is now more compliant with the spec which means some code that may have previously worked (or failed silently) will now cause an error.

Migration steps:
The content string must be converted to a `java.nio.ByteBuffer` instead of setting it directly to the request options. Here's an example of converting a content string:
```ts
        const contentString = '...blah...';
        const nativeString = new java.lang.String(contentString);
        const nativeBytes = nativeString.getBytes('UTF-8');
        const nativeBuffer = java.nio.ByteBuffer.wrap(nativeBytes);
        myhttpRequestOptions.content = nativeBuffer;
```

rynop commented:
Thanks @charsleysa , really appreciate this.  Few things:

- How are you handling `application/protobuf` (binary) responses? AFAIK NS does not currently support (I think confirmed by the https://github.com/NativeScript/nativescript-angular/issues/1202 you mention)
-  Do you have any examples/testcases? From looking at code should just need to set `content` to an `ArrayBuffer` however having something concrete to start from would save me time.  Sounds like you use Angular - Angular ex would be perfect.

FWIW I'm willing to test and write the docs

charsleysa commented:
Hi @rynop 

Regarding your first point, you are correct that binary download support is not currently available through angular due to XHR not having binary response support. I will add in arraybuffer responseType support to XHR and include it in this PR.

As for your second point, any angular tutorial for arraybuffer should work. Here's the http snippet from our app when we upload a user profile image:
```ts
                            this.http
                                .request(
                                    new HttpRequest('PUT', uploadUrl, data, {
                                        headers: new HttpHeaders({
                                            'Content-Type': fileType,
                                            'Content-Disposition': `attachment; filename="${fileName}"`
                                        })
                                    })
                                )
```


If you're able to write tests and docs that'd be a great help!

charsleysa commented:
@rynop I've implemented full support (within limitations) for Blob and ArrayBuffer in both upload and download for XHR, updated the Fetch API polyfill. I've also added some tests but this PR will probably need a bit of testing against a few apps to make sure nothing is horribly broken.

rynop commented on tests/app/fetch/fetch-tests.ts:
```diff
@@ -54,6 +54,30 @@ export var test_fetch_formData = function (done: (err: Error, res?: string) => v
+export var test_fetch_blob = function (done: (err: Error, res?: string) => void) {
+    // >> fetch-blob
+    fetch("https://httpbin.org/get").then(response => response.blob()).then(function (r) {
+        // Argument (r) is Blob object!
```
https://httpbin.org/get doesn't return binary data correct? Wondering if this is a valid test. I know the JSON response is Gzipped, but I'm concerned its deflated via https://github.com/NativeScript/NativeScript/issues/1314  

Or are you just testing that you can convert a string response from fetch to a blob?

charsleysa replied:
This is just testing the Fetch API to see if it will return the proper type.

rynop commented on tests/app/fetch/fetch-tests.ts:
```diff
@@ -54,6 +54,30 @@ export var test_fetch_formData = function (done: (err: Error, res?: string) => v
+export var test_fetch_arraybuffer = function (done: (err: Error, res?: string) => void) {
+    // >> fetch-arraybuffer
+    fetch("https://httpbin.org/get").then(response => response.arrayBuffer()).then(function (r) {
+        // Argument (r) is ArrayBuffer object!
```
Same here, https://httpbin.org/get returns Gzipped string (JSON) not binary data...

charsleysa replied:
Same as above.

rynop commented on tns-core-modules/http/http.ts:
```diff
@@ -58,3 +58,17 @@ export function getFile(arg: any, destinationFilePath?: string): Promise<any> {
             }, e => reject(e));
     });
 }
+
```
Need to support more than the `GET` method.  For example, `POST`ing protobuf and getting a RPC protobuf (binary) response is very typical.

I think ideally all the existing NS http module [request](https://docs.nativescript.org/ns-framework-modules/http) methods would inspect the response determine if its is binary automatically.

Said another way, when using XHR/fetch from browser I don't have to know ahead of time if the response will be string or not - I just get back a string or an ArrayBuffer.  If this isn't possible, we'd need `[post|put|..]Binary` methods as well.

It looks like my outlined logic IS what you are doing in your new changes to tns-core-modules/xhr/xhr.ts.  If so I'm confused why you need this `getBinary()` method.  So I'm probably missing something here....

charsleysa replied:
Http lib is only a basic API. If you want full support you must use Fetch of XHR. Also, as per spec for XHR you must specify ahead of time what type of response you want, XHR may not change the response type. You can get close to the functionality you are after by using the Fetch API but it's still not quite the same. Both XHR and Fetch aim to meet spec, the http lib is an underlying platform abstraction which I don't recommend using directly.

rynop replied:
Makes sense, thanks.  I like you, am a big fan of Angular. You recommend staying away from any abstraction? Your [snippet above](https://github.com/NativeScript/NativeScript/pull/7707#issuecomment-523660311) is using the [@angular/common/http](https://angular.io/api/common/http) abstraction of `XMLHttpRequest` correct? This abstraction safe to use in NS in your experience (I'll obv test)?

charsleysa replied:
Angular HttpClient uses XHR under the hood so Angular's abstraction should work completely fine. One note (which if you go through the code you'll be able to see) is that the lower level platform abstractions don't provide progress reports so XHR only reports progress once on completion of the request.

charsleysa commented:
@rynop did you solve your problem and get this running?
I have not tested linking the repo to another NS app project.

rynop commented:
@charsleysa nope, I think there is a `tns-core-modules` bug in master (and therefore your fork/branch).  Reproduction of bug and tracking of it is in "Update DevelopmentWorkflow.md for v6" link above.

bundyo commented:
The problem with linking is that Angular/TS project tend to grab the linked TS files in core modules and try to compile them without the references needed. I've only managed to run it by compiling the core modules with outDir and linking that (.d.ts, package.json and other files, except TS ones should be copied over too). Maybe the team can consider moving to an outDir compilation. :)

rynop commented:
@charsleysa It was a long (30+) day journey but I _finally_ was able to verify the changes in the PR work great when using binary protobuf payloads.  Can't tell you how much I appreciate this PR, I really think it is a big step forward for NS and gives a competitive advantage over React Native (which last time I checked has problems passing binary data over their JS bridge).

Now the details.

**Workaround so I could test:**
[As suspected](https://github.com/NativeScript/NativeScript/issues/7745), the [development workflow](https://github.com/NativeScript/NativeScript/blob/master/DevelopmentWorkflow.md) for contributing to `tns-core-modules` and `tns-platform-declarations` do not work when verifying using an Angular flavor NS app.  After lots of investigation and debugging I have documented a workaround in https://github.com/NativeScript/NativeScript/pull/7861 (see `files changed`).  It is nasty and the NS team agrees.  They are working on a better approach.

**What I tested**
Binary support you've added fetch/XHR is awesome.  No longer limited to base64 encode a [background thread](https://github.com/NativeScript/nativescript-background-http) which both have obv drawbacks.  My main drive for binary support was to [get protobuf working in NS](https://github.com/NativeScript/NativeScript/issues/7632).  

In my test, I ran against a [Twirp RPC](https://github.com/twitchtv/twirp) server.  For those curious, I used https://github.com/rynop/abp-sam-twirp almost verbatim.

I then created an Angular NS project, and an Angular service with the method below, which invokes the code in this PR.

```
  public createTwirpRPCAdapter(servicePathPrefix: string): $protobuf.RPCImpl {
    return (serviceMethodName: $protobuf.Method, requestData: Uint8Array, callback: Function) => {
      const req = new HttpRequest(
        'POST',
        servicePathPrefix + TwirpService.getServiceMethodName(serviceMethodName),
        // required to get an arraybuffer of the actual size, not the 8192 buffer pool that protobuf.js uses
        // see: https://github.com/dcodeIO/protobuf.js/issues/852
        requestData.slice().buffer,
        {
          headers: new HttpHeaders({
            'Accept': 'application/protobuf',
            'Content-Type': 'application/protobuf'
          }),
          responseType: 'arraybuffer',
        }
      );

      this.http.request<ArrayBuffer>(req)
        .subscribe(
          event => {
            console.log('event', event);
            if (event instanceof HttpResponse) {
              callback(null, new Uint8Array(event.body));
            }
          },
          (error: HttpErrorResponse) => {
            return callback(TwirpError.isTwirpError(error) ? new TwirpError(error) : error, null);
          }
        );
    };
  }
```

I successfully tested both green and error (500/404/400 etc) paths.

There is quite a bit more to the protobuf aspect, but I don't want to bloat this PR.  I'll blog and/or post more details for others wanting this use case in https://github.com/NativeScript/NativeScript/issues/7632

@bundyo the only issue I see is the potentially breaking change (`org.nativescript.widgets2.Async.Http`) mentioned above.  Is this considered an internal API? If not do you know someone at Telerik / maintainer that can make the call? As stated above I think this PR has some significant value, so would hate to see it get held up too long.

bundyo commented:
@manoldonev Any chance this hitting 6.2?

JoeyBG commented:
We really need this as well! Simply trying to upload a raw image to S3..

manoldonev commented:
@bundyo most probably will not ship with 6.2 but this PR is definitely on our radar and we will do our best to process it as soon as possible (and release it officially with 6.3)

rynop commented:
I should have noted that my test above uses `HttpRequest` from [@angular/common/http](https://angular.io/api/common/http) which under the covers uses `XMLHttpRequest`.

For completeness sake, I just tested `POST`ing protobuf via [fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch) like so:

```javascript
  public createTwirpRPCAdapter(servicePathPrefix: string, headerOverrides?: HttpHeaders): $protobuf.RPCImpl {
    return (serviceMethodName: $protobuf.Method, requestData: Uint8Array, callback: Function) => {
      fetch(servicePathPrefix + TwirpService.getServiceMethodName(serviceMethodName),
        {
          method: 'POST',
          headers: {
            'Accept': 'application/protobuf',
            'Content-Type': 'application/protobuf'
          },
          body: requestData.slice().buffer,
        })
        .then(response => response.arrayBuffer())
        .then(body => callback(null, new Uint8Array(body)))
        .catch(error => {
          callback(TwirpError.isTwirpError(error) ? new TwirpError(error) : error, null);
        });
    };
  }
```

Sending binary in the request body and getting binary in the response body works in my tests via `XMLHttpRequest` and `fetch`.

acivier-serial commented:
Hello!
Could you confirm that this feature will be present in the next release? (6.3)

rynop commented:
@manoldonev since this did not make it into v6.3 as hoped, can you confirm it is slated to go into v6.4?

vahidvdn commented:
Is the PR stuck for some reason?

rynop commented:
From my perspective it is ready to go.  I tested multiple scenarios on multiple devices/emulators/simulators.   I'm guessing the team wants to get more testing (wise).

vtrifonov commented:
test

vtrifonov commented:
test

vtrifonov commented:
test

vtrifonov commented:
test

rynop commented:
Wahooo!!! Thanks for everyone's hard work on this! Especially you @charsleysa

vibos commented:
I'm trying to upload image in binary mode using Angular's `HttpClient`. I tried different options, also tried using fetch and xhr, but nothing works. 
I have `ImageSource` as input, then saving it to filesystem and need to send POST request with `multipart/form-data` type.
Could anyone please provide code snippet how to do that?
